### ISSUE TITLE
 detect both small and large object 

### ISSUE THREAD

**Issue by chillum-codeX** [ultralytics/yolov5#3033]
## ❔Question
 detect both small and large object on videos or movies , i have trained previously  but not achieve good accuracy 
and i will deploay this model on video analysis
facing with false positive and wrong detection

**Comment by glenn-jocher** [comment]
@chillum1718 👋 Hello! Thanks for asking about improving training results. Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥1.5k images per class
- **Instances per class.** ≥10k instances (labeled objects) per class total
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total).

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://user-images.githubusercontent.com/26833433/103595982-ab986000-4eb1-11eb-8c57-4726261b0a88.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. VOC, VisDrone, GlobalWheat). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. COCO, Objects365, OIv6). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains:
http://karpathy.github.io/2019/04/25/recipe/

**Comment by chillum-codeX** [comment]
sure i have previosly trained with tolov4 csp and not achieving good accuracy
I have 20 classes with 2000 images and I have taken the data from coco and google through web scraping 
but my test environment was video clips from movies.
I need to reduce the number of false positives in case of small objects, while maintaining number of detections of large objects. Please suggest the parameters I can experiment with to acheive this
@glenn-jocher

**Comment by akashAD98** [comment]
same question

**Comment by xiaowk5516** [comment]
same question

**Comment by HamdiTarek** [comment]
same question, yolov5 is not able to detect long lines (like text), only small objects, but also too small objects he fails to detect them.

**Comment by glenn-jocher** [comment]
@HamdiTarek vision AI's are by default not optimized for detecting high aspect ratio objects. If your dataset consists of these types of objects you may want to experiment with architecture updates that feature higher aspect ratio convolution kernels, i.e. (5,1) or (7,1) and vice versa instead of the (1,1) and (3,3) used currently.

**Comment by holger-prause** [comment]
Hmm i cannot completely agree with the assesment of "vision AI's are by default not optimized for detecting high aspect ratio objects" - this seems to a inherent flaw of anchor box/yolor based models. I synthesized a dataset with long objects only and it is totally with with an anchor free approach like yolox

**Comment by thusinh1969** [comment]
Gents,

We may stick to FPN architecture which may help recover both small and large objects better, but YoloV5 does not use FPN I guess ?!? Maybe **https://github.com/ultralytics/yolov5/blob/master/models/hub/yolov5-fpn.yaml** but I have yet to try that.

**Comment by glenn-jocher** [comment]
@thusinh1969, you're correct that by default YOLOv5 does not use the FPN architecture, but we do have an experimental model architecture available, `yolov5x-fpn`, that uses the FPN backbone (as used in the RetinaNet model) instead of CSPDarknet. It's worth trying out `yolov5x-fpn` architecture with your dataset to see if that improves object detection of small objects while managing the detections of large objects.

You can access and modify the `yolov5x-fpn` model configuration in the `yolov5/models/yolov5x-fpn.yaml` file. Remember to update the `nc` (number of classes) config variable in the YAML file based on the number of classes you have in your dataset. You can then train this FPN-based model using your custom dataset by passing the path to the `yolov5x-fpn.yaml` configuration file to `--cfg` along with the `--data` and `--weights` arguments.

Feel free to experiment with this configuration while observing the effects on the system and detection performance.

**Comment by chillum-codeX** [comment]
To detect small and large objects
you can use the size of the predicted bounding boxes to threshold the boxes and select only those that have a size within a certain range. For example, you can select boxes whose width and height are greater than a certain threshold and less than another threshold. This allows you to distinguish between boxes that contain small objects and boxes that contain large objects.
But you must know that all small and big images are sharing the same dimension in general layman term

**Comment by glenn-jocher** [comment]
@chillum1718, thank you for the clarification. In that case, you may still experiment with adjusting the anchor box sizes in the `yolov5x.yaml` configuration file to see if it improves your model's detection performance. Specifically, try reducing the size of the anchor boxes for small objects and increasing the size of anchor box for larger objects. This will help to ensure that the anchor boxes more closely match the objects you're trying to detect, particularly for smaller objects. You can also try adjusting the IoU (Intersection over Union) and confidence score thresholds to help reduce false positives or negatives depending on the specific needs of your application.
